crates/codegen/schema/src/types/precedence_parser.rs

```diff
@@ -0,0 +1,51 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use super::parser::ParserDefinition;
+
+#[allow(dead_code)]
+pub type PrecedenceParserRef = std::rc::Rc<PrecedenceParser>;
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq, Hash)]
+#[serde(deny_unknown_fields)]
+pub struct PrecedenceParser {
+    #[serde(default)]
+    pub name: Option<String>,
+    #[serde(flatten)]
+    pub definition: PrecedenceParserDefinition,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq, Hash)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct PrecedenceParserDefinition {
+    #[schemars(title = "Operator Definitions")]
+    pub operators: Vec<OperatorDefinition>,
+
+    #[schemars(title = "Primary Expression References")]
+    pub primary_expressions: Vec<Reference>,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq, Hash)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct OperatorDefinition {
+    pub name: String,
+    pub model: OperatorModel,
+    #[serde(flatten)]
+    pub definition: ParserDefinition,
```

Comment:
```suggestion
    pub operator: ParserRef,
```

In the current grammar, `name` here is the name of the parent expression node (that contains both the operator and operands). There is no way to define a name for the operator node.

```yaml
        - name: "UnarySuffixExpression" # name of parent node
          model: "UnarySuffix"
          choice: # cannot name the operator node
            - terminal: "++"
            - terminal: "--"
```

I suggest the following changes:

1. Not flattening it.
2. Changing the field name to `operator` to indicate that this is the operator definition, not the parent.
3. Change its type to `ParserRef`, to allow us to specify names.


```yaml
        - name: "UnarySuffixExpression" # name of parent node
          model: "UnarySuffix"
          operator:
            name: "Operator" # name of the operator node
            choice:
                - terminal: "++"
                - terminal: "--"
```

Reply:
I think because of this, they are also excluded from the `visit_parser` check, which is a bug.

Reply:
Will create an issue for that, so that we can reference to the actual token productions from there

Reply:
* https://github.com/NomicFoundation/slang/issues/374